### PR TITLE
First And Last Names Tests Now Pass. Also, added more tests to First and Last Names.

### PR DESCRIPTION
**Pull request by vardahoth**
I added some First and Last Names tests, and filled in the code to make all these First and Last Names Tests to pass. I would like to discuss any needed changes, before I get the other property tests to pass.

### CODE REVIEW

File: src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java [hunk 1]
```diff
@@ -8,18 +11,60 @@
...
+
     private static AnswerType getMatchingResultFromMatchingMethodsList(String contactOneFirstName, String contactOneLastName, String contactTwoFirstName, String contactTwoLastName){
         // Do first and last names have an exact match (case sensitive)? If so, return yes.
-        if (isFirstAndLastNamesExactMatchCaseSensitive(contactOneFirstName, contactOneLastName, contactTwoFirstName, contactTwoLastName)){
+        if (isMatchMatchCaseSensitive(contactOneFirstName, contactOneLastName, contactTwoFirstName, contactTwoLastName)){
             return AnswerType.yes;
         }
 
         // Do first and last names have an exact match (case Insensitive)? If so, return yes.
+        if (isMatchCaseInsensitive(contactOneFirstName, contactOneLastName, contactTwoFirstName, contactTwoLastName)){
```
Comment (mbilodeau6) [inline src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java]: Doing a separate test for isMatchCaseInsensitive() and isMatchWithWhiteSpaces() is unnecessary. They are already accomplished by your isMatchWithWhiteSpaces_CaseInsensitive(). An exact match with spaces and case checking isn't "more" AnswerType.yes. In fact, I think you should even be able to perform all of them with a call to isMatchWithPunctuations() if you modify isMatchWithPunctuations() to do a case insensitive compare. Less is more.

File: src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java [hunk 2]
```diff
@@ -28,13 +73,145 @@ private static AnswerType getMatchingResultFromMatchingMethodsList(String contac
...
-        return AnswerType.no;
+    private static boolean isMatchWithWhiteSpaces(String contactOneFirstName, String contactOneLastName, String contactTwoFirstName, String contactTwoLastName){
+        return (contactOneFirstName.trim().equals(contactTwoFirstName.trim()) && contactOneLastName.trim().equals(contactTwoLastName.trim()));
+    }
+
+    private static boolean isMatchWithWhiteSpaces_CaseInsensitive(String contactOneFirstName, String contactOneLastName, String contactTwoFirstName, String contactTwoLastName){
+        return (contactOneFirstName.trim().equalsIgnoreCase(contactTwoFirstName.trim()) && contactOneLastName.trim().equalsIgnoreCase(contactTwoLastName.trim()));
+    }
+
+    private static boolean isMatchWithPunctuations(String contactOneFirstName, String contactOneLastName, String contactTwoFirstName, String contactTwoLastName){
```
Comment (mbilodeau6) [inline src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java]: Should be "isMatchWithoutPunctuation()".

File: src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java [hunk 3]
```diff
@@ -28,13 +73,145 @@ private static AnswerType getMatchingResultFromMatchingMethodsList(String contac
...
+    private static String replacePunctuationsInNameWithWhiteSpace(String name){
+        String nameWithoutPunctuations = name;
+
+        for (String punctuation: namePunctuations){
+            nameWithoutPunctuations = nameWithoutPunctuations.replace(punctuation, " ");
+        }
+        return nameWithoutPunctuations;
+    }
+
+    private static String removePunctuationsInName(String name){
```
Comment (mbilodeau6) [inline src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java]: The code in this method is identical to the code in replacePunctuationsInNameWithWhiteSpace(). I don't think we need this version as the test with whitespace removed is the one we want to use all the time anyways.

File: src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java [hunk 4]
```diff
@@ -28,13 +73,145 @@ private static AnswerType getMatchingResultFromMatchingMethodsList(String contac
...
+            if (contactTwoFirstNameWithoutPunctuations.startsWith(contactOneFirstNameWithoutPunctuations)){
+                firstNameIsInitialOf2ndFirstName = true;
+            }
+        } else if (isNameInitial(contactTwoFirstNameWithoutPunctuations)){
+            if (contactOneFirstNameWithoutPunctuations.startsWith(contactTwoFirstNameWithoutPunctuations)){
+                firstNameIsInitialOf2ndFirstName = true;
+            }
+        }
+
+        // Last name is an initial
```
Comment (mbilodeau6) [inline src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java]: Last name can never match if it is just an initial. I don't even think we can call that one a maybe. We should add a test to ensure that we never declare a name match in this case as I think it would result in a bunch of bad recommendations.

File: src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java (hunk 3 above)
Comment (mbilodeau6) [reply src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java]: Ah... I see the difference now. One you replace with a space and another with an empty string. Do we really need both? If yes, still should refactor to reduce duplicate code. One way would be to create another method that accepts the replacement character and then each of the existing methods would call with the appropriate replacement character.

File: src/com/cft/contactmerge/tests/CompareContactPartsTest.java [hunk 5]
```diff
@@ -6,176 +6,200 @@
...
 
     @Test
-    void doNamesMatch_MaybeFirstNameInitials() {
+    void doNamesMatch_Maybe_FirstNameInitials() {
         assertEquals(AnswerType.maybe, CompareContactParts.doNamesMatch("John", "Doe", "J", "Doe"));
         assertEquals(AnswerType.maybe, CompareContactParts.doNamesMatch("J.", "Doe", "John", "Doe"));
     }
 
+    @Test
+    void doNamesMatch_no_FirstNameInitialsDifferentLastName() {
```
[the remark this answers is not in the document]
Comment (vardahoth) [reply src/com/cft/contactmerge/tests/CompareContactPartsTest.java]: Thanks for the catch. Will change this and re-upload.

File: src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java (hunk 1 above)
Comment (vardahoth) [reply src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java]: Yup, as discussed yesterday, I am going to refactor the code to use very little methods. I will use a string formatted method to change strings for matching, so very little methods will be used.

File: src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java (hunk 2 above)
Comment (vardahoth) [reply src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java]: I'm a little confused by this request... If the method is given a name "with" a punctuation (. , - _ ') in it... wouldn't it be trying to match against a name "with/without" punctuations?
vs
Given all names "without" punctuation to match? (in which case this method would never be called)

Comment (vardahoth) [reply src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java]: ah I get what you mean.... if punctuations are removed (names "without" punctuations)... do the names match. Okay I can make that change.

File: src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java (hunk 3 above)
Comment (vardahoth) [reply src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java]: I agree with passing the replacement character as an argument to remove duplicate code. When I refactor the code, I may not need both. I was thinking in terms of handling separate cases. But if we are set on a generic matching logic, I probably will only be replacing with white spaces. I will have to see....

File: src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java (hunk 4 above)
Comment (vardahoth) [reply src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java]: I'm going to give a case here: When I was growing up with someone, he always was called by his initials TJ. So in two database he might be known as (firstName: T | lastName: J) and (firstname: Tom | lastName: Jacobs).

That being said, confirming you don't think the match should be a maybe, because the last name is an initial?

File: src/com/cft/contactmerge/tests/CompareContactPartsTest.java [hunk 6]
```diff
@@ -6,176 +6,200 @@
...
+    void doNamesMatch_Yes_IgnorePunctuation() {
         assertEquals(AnswerType.yes, CompareContactParts.doNamesMatch("Jane", "Doe-Adams", "Jane", "Doe Adams"));
     }
 
     @Test
-    void doNamesMatch_YesHyphenatedNames() {
-        assertEquals(AnswerType.yes, CompareContactParts.doNamesMatch("Jane", "Doe-Adams", "Jane", "Doe"));
-        assertEquals(AnswerType.yes, CompareContactParts.doNamesMatch("Jane", "Adams", "Jane", "Doe-Adams"));
+    void doNamesMatch_Maybe_HyphenatedNames() {
+        assertEquals(AnswerType.maybe, CompareContactParts.doNamesMatch("Jane", "Doe-Adams", "Jane", "Doe"));
```
[the remark this answers is not in the document]
Comment (vardahoth) [reply src/com/cft/contactmerge/tests/CompareContactPartsTest.java]: Giving a what if case: I was thinking what if the we have a case of two different last names with hyphens containing the same first part. For example... ("la-chan" and "la-shin") or ("chan-la" and "shin-la"). Two separate people, but because they both have "la" in them, the code would find a successful match.

Since there is a medium possibility of the full names not matching, shouldn't this be a maybe?

**Comment by vardahoth** [comment]
There were two comments I disagree'd with, and we can have a discussion about it if you want. The rest look great and I appreciate the feedback. I'll wait till we agree on both comments before checking it in. Still working on the updated code though.




Thanks,

-James


________________________________

File: src/com/cft/contactmerge/tests/CompareContactPartsTest.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (mbilodeau6) [reply src/com/cft/contactmerge/tests/CompareContactPartsTest.java]: The code would say that there is a successful match on the last name, not on the whole person.  Forget about hyphenated names... There are many people with the last name "Page" or "Bilodeau".  When we recommend matches we check all properties to try to determine if it is the same person. If the name and address are the same we can be pretty sure (but not 100% due to Jr/Sr) that it is the same person. The other thing that protects us is that if we get "match" on more than 1 contact then we can let a human pick which one to use. As I put in my original comment, I won't be surprised if we discover exceptions at some point that will force us to tweak this rule but I'd like to have some real examples before we make the code more complicated. I think it could be reasonable to say that there is some size limit for when we will match on partial last name. For example, if we find multiple names in a last name, matches on parts that are 3 or less characters result in a maybe while matches of names with over 3 characters result in a yes?

File: src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java (hunk 4 above)
Comment (mbilodeau6) [reply src/com/cft/contactmerge/FirstAndLastNamesMatchLogic.java]: If it is an exact match on the last name, we should have it match. But I wouldn't do the "does the first letter match the initial" on last name. For example, where C1 is Contact1 and C2 is Contact2, and F is First, L is Last.... C1F: T; C1L: J should have a "maybe match" on C2F:Tommy, C2L:J, a "match" on C2F:T, C2L:J, and a "no match" on C2F:T, C2L:Johnson. We can discuss more in the call but if you agree, can you add these tests before making the change?
